### PR TITLE
New package: QGIS

### PR DESCRIPTION
**Pull request by Sinan81**
Work in progress. 
- ~~one missing dependency: qtkeychain (not sure if this is a critical dependency)~~
- ~~dependency Qscintilla doesn't yet compile with newer versions of pyqt@4.12.1:, py-pyqt needs to be compiled with a private SIP module~~
- ~~dependency QCA requires a newer version (2.2.1) to compile with Qt@5.11.3, also needs to be compiled with CMAKE_CXX_STANDARD=11~~

### CODE REVIEW

**Comment by Sinan81** [comment]
@adamjstewart : I figured I should mention you here.

**Comment by adamjstewart** [comment]
Feel free to add me as a maintainer. I use QGIS pretty regularly and it'd be nice to build a copy that supports more file formats than the defaults. Now I'll have to figure out how to register the installation as an app with a dock icon.

**Comment by Sinan81** [comment]
Just added you as a maintainer. It is lucky for me that you, one of main spack guys, are working in the area of remote sensing!

**Comment by Sinan81** [comment]
So, have you managed to compile it yet? I am still stuck at compiling Qscintilla python bindings which seemed to require pyqt to be compiled with private sip module as suggested in #12353 . I implemented what was suggested in the PR but still getting 'PyQt5.sip' not found error. #12280 

I guess for now I will proceed with skipping python bindings, and see what happens.

**Comment by adamjstewart** [comment]
I haven't tried to compile it yet, but I will when this is no longer a WIP. I'll try to take another look at #12353 and #12280. If I forget, remind me in a few days. I'm moving aprts this weekend.

**Comment by Sinan81** [comment]
Managed to build qscintilla python bindings using a symlink hack. Afterwards did the first attempt of building qgis where noteworthy parts are:
```
...
-- Found OpenCL: /usr/lib64/libOpenCL.so (found version "2.2") 
-- Found OpenCL C++ headers: /usr/include
-- Could not find GRASS 7
...
-- Qt WebKit support DISABLED.
...
-- Found PyQt5 version: 5.13.0
-- Found SIP version: 4.19.18
Traceback (most recent call last):
  File "/cache/spack-stage/sbulut/qgis-3.8.1-ifxneplvshnn2hyvs5h5c6oyyqn6ub7m/spack-src/cmake/FindQsci.py", line 45, in <module>
    from PyQt5.Qsci import QSCINTILLA_VERSION_STR
ModuleNotFoundError: No module named 'PyQt5.Qsci'
-- txt2tags not found - disabled
...
-- Found PkgConfig: /usr/bin/pkg-config (found version "0.27.1") 
-- Could NOT find NetCDF (missing: NETCDF_LIBRARY NETCDF_INCLUDE_DIR) 
...
**-- Using PROJ <6 srs database.**
-- Qsci sip file not found - disabling bindings for derived classes
...
-- Configuring done
WARNING: Target "qgis_gui" requests linking to directory "/mnt/local/spack/opt/spack/linux-centos7-x86_64/gcc-8.2.0/qscintilla-2.10.2-jxz7c7iyrldhfr2fiyhlyvf6dq4podoo/lib".  Targets may link only to libraries.  CMake is dropping the item.
WARNING: Target "basicauthmethod" requests linking to directory "/mnt/local/spack/opt/spack/linux-centos7-x86_64/gcc-8.2.0/qscintilla-2.10.2-jxz7c7iyrldhfr2fiyhlyvf6dq4podoo/lib".  Targets may link only to libraries.  CMake is dropping the item.
...
```

**Comment by Sinan81** [comment]
builds fails due to `zlibconf.h` not found error. Tried adding the corresponding include path to cmake args via `-DLIBZIP_CONF_INCLUDE_DIR` which didn't work. Proceded with another symlink hack:
```
ln -s /path/to/libzip../lib/libzip/include/zlibconf.h /path/to/libzip../include/zlibconf.h

**Comment by Sinan81** [comment]
Next attempt: build progresses 25%, yielding a sip error: it seems build process can't locate '*.sip' files which are located under python prefix as opposed to py-sip or py-pyqt5!
```
     3748    [ 25%] Generating flex_qgsrastercalclexer.cpp
     3749    cd /mnt/local/cache/sbulut/spack-stage/qgis-3.8.1-6levp6gscjhevnma
             rr5zwd3m3vizjd3d/spack-build/src/analysis && /disk/software/lib/hp
             cpm/spack/opt/spack/linux-centos7-x86_64/gcc-7.4.0/flex-2.6.4-4i5j
             ekvrmxofj2xuqwmfie5l6zfdvxjc/bin/flex -Praster -o/mnt/local/cache/
             sbulut/spack-stage/qgis-3.8.1-6levp6gscjhevnmarr5zwd3m3vizjd3d/spa
             ck-build/src/analysis/flex_qgsrastercalclexer.cpp /cache/sbulut/sp
             ack-stage/qgis-3.8.1-6levp6gscjhevnmarr5zwd3m3vizjd3d/spack-src/sr
             c/analysis/raster/qgsrastercalclexer.ll
     3750    cd /mnt/local/cache/sbulut/spack-stage/qgis-3.8.1-6levp6gscjhevnma
             rr5zwd3m3vizjd3d/spack-build && /disk/software/lib/hpcpm/spack/opt
             /spack/linux-centos7-x86_64/gcc-7.4.0/cmake-3.15.1-gxyqb6bcuppyki7
             tajwhbfchkwarurjf/bin/cmake -E cmake_depends "Unix Makefiles" /cac
             he/sbulut/spack-stage/qgis-3.8.1-6levp6gscjhevnmarr5zwd3m3vizjd3d/
             spack-src /cache/sbulut/spack-stage/qgis-3.8.1-6levp6gscjhevnmarr5
             zwd3m3vizjd3d/spack-src/src/providers/gpx /mnt/local/cache/sbulut/
             spack-stage/qgis-3.8.1-6levp6gscjhevnmarr5zwd3m3vizjd3d/spack-buil
             d /mnt/local/cache/sbulut/spack-stage/qgis-3.8.1-6levp6gscjhevnmar
             r5zwd3m3vizjd3d/spack-build/src/providers/gpx /mnt/local/cache/sbu
             lut/spack-stage/qgis-3.8.1-6levp6gscjhevnmarr5zwd3m3vizjd3d/spack-
             build/src/providers/gpx/CMakeFiles/gpxprovider.dir/DependInfo.cmak
             e --color=
     3751    sip: Unable to find file "QtXml/QtXmlmod.sip"
  >> 3752    make[2]: *** [python/core/sip_corepart0.cpp] Error 1
     3753    make[2]: *** Deleting file `python/core/sip_corepart0.cpp'
     3754    make[2]: Leaving directory `/mnt/local/cache/sbulut/spack-sta
```

**Comment by Sinan81** [comment]
Need to check out a file that the cmake option `_find_sip.py` is pointing to:
```
/path/to/stage/qgis-3.8.1-6levp6gscjhevnmarr5zwd3m3vizjd3d/spack-src/cmake/FindSIP.py
```
A possible hackish solution is mentioned at https://issues.qgis.org/issues/16034

**Comment by Sinan81** [comment]
output from FindSIP.py is a follows. Note that 'default sip dir ../share/sip' doesn't exist
```
$ python FindSIP.py
sip_version:041312
sip_version_num:267026
sip_version_str:4.19.18
sip_bin:/disk/software/lib/hpcpm/spack/opt/spack/linux-centos7-x86_64/gcc-7.4.0/py-sip-4.19.18-jripvr46bqmsscbq4vaszjorgrxd6i3b/bin/sip
default_sip_dir:/disk/software/lib/hpcpm/spack/opt/spack/linux-centos7-x86_64/gcc-7.4.0/py-sip-4.19.18-jripvr46bqmsscbq4vaszjorgrxd6i3b/share/sip
sip_inc_dir:/disk/software/lib/hpcpm/spack/opt/spack/linux-centos7-x86_64/gcc-7.4.0/py-sip-4.19.18-jripvr46bqmsscbq4vaszjorgrxd6i3b/include/python3.7
sip_module_dir:/disk/software/lib/hpcpm/spack/opt/spack/linux-centos7-x86_64/gcc-7.4.0/py-sip-4.19.18-jripvr46bqmsscbq4vaszjorgrxd6i3b/lib/python3.7/site-packages
```
Also
```
$ module list
Currently Loaded Modulefiles:
  1) python-3.7.4-gcc-7.4.0-mbxbk5x        3) qscintilla-2.10.2-gcc-7.4.0-pc3qvdj   5) py-sip-4.19.18-gcc-7.4.0-jripvr4
  2) py-pyqt5-5.13.0-gcc-7.4.0-z43mdws     4) gcc-5.5.0-gcc-7.4.0-2zwfggb
```

**Comment by Sinan81** [comment]
Used another symlink hack, now **compilation completes 61%**
```
mkdir path/to/py-sip-4.19.18-jripvr46bqmsscbq4vaszjorgrxd6i3b/share 
ln -s /path/to/python-3.7.4-mbxbk5xl34brqfl6542u3ratm456rmkl/share/sip /path/to/py-sip-4.19.18-jripvr46bqmsscbq4vaszjorgrxd6i3b/share/sip
```
Afterwards tried compiling
```
==> Searching for binary cache of qgis
==> Warning: No Spack mirrors are currently configured
==> No binary for qgis found: installing from source
==> Using cached archive: /disk/software/lib/hpcpm/spack/var/spack/cache/qgis/qgis-3.8.1.tar.bz2
==> Staging archive: /cache/sbulut/spack-stage/qgis-3.8.1-6levp6gscjhevnmarr5zwd3m3vizjd3d/qgis-3.8.1.tar.bz2
==> Created stage in /cache/sbulut/spack-stage/qgis-3.8.1-6levp6gscjhevnmarr5zwd3m3vizjd3d
==> No patches needed for qgis
==> Building qgis [CMakePackage]
==> Executing phase: 'cmake'
==> Executing phase: 'build'
==> Error: ProcessError: Command exited with status 2:
    'make' '-j16'

872 errors found in build log:
     10852    [ 61%] Building CXX object src/gui/CMakeFiles/qgis_gui.dir/editor
              widgets/moc_qgsqmlwidgetwrapper.cpp.o
     10853    cd /mnt/local/cache/sbulut/spack-stage/qgis-3.8.1-6levp6gscjhevnm
              arr5zwd3m3vizjd3d/spack-build/src/gui && /disk/software/lib/hpcpm
...
     10857    CMakeFiles/qgis_gui.dir/qgscodeeditor.cpp.o: In function `QgsCode
              Editor::insertText(QString const&)':
  >> 10858    /cache/sbulut/spack-stage/qgis-3.8.1-6levp6gscjhevnmarr5zwd3m3viz
              jd3d/spack-src/src/gui/qgscodeeditor.cpp:182: undefined reference
               to `QsciScintilla::getCursorPosition(int*, int*) const'
     10859    CMakeFiles/qgis_gui.dir/qgscodeeditor.cpp.o: In function `QgsCode
              Editor::keyPressEvent(QKeyEvent*)':
  >> 10860    /cache/sbulut/spack-stage/qgis-3.8.1-6levp6gscjhevnmarr5zwd3m3viz
              jd3d/spack-src/src/gui/qgscodeeditor.cpp:82: undefined reference 
              to `QsciScintilla::isListActive() const'
...
```

**Comment by Sinan81** [comment]
Turns out I didn't actually build Qsci python bindings rather just configured it, generating a Makefile.
This should take care of import error for 'PyQt5.Qsci'.

**Comment by Sinan81** [comment]
As of today, Qscintilla python bindings build just fine as discussed in the corresponding issue.

**Comment by Sinan81** [comment]
Regardless of python bindings, QGIS compilation gets stuck at 61%. There seems to be linking issue with QGIS.

**Comment by Sinan81** [comment]
cmake output and spack build output has clues about what the problem is:
```
...
-- Configuring done
WARNING: Target "qgis_gui" requests linking to directory "/disk/software/lib/hpcpm/spack/opt/spack/linux-centos7-x86_64/gcc-7.4.0/qscintilla-2.10.2-bdxsvd7op5vk2d6pnf6baghbsssghhyz/lib".  Targets may link only to libraries.  CMake is dropping the item.
WARNING: Target "basicauthmethod" requests linking to directory "/disk/software/lib/hpcpm/spack/opt/spack/linux-centos7-x86_64/gcc-7.4.0/qscintilla-2.10.2-bdxsvd7op5vk2d6pnf6baghbsssghhyz/lib".  Targets may link only to libraries.  CMake is dropping the item.
WARNING: Target "esritokenauthmethod" requests linking to directory "/disk/software/lib/hpcpm/spack/opt/spack/linux-centos7-x86_64/gcc-7.4.0/qscintilla-2.10.2-bdxsvd7op5vk2d6pnf6baghbsssghhyz/lib".  Targets may link only to libraries.  CMake is dropping the item.
...
```

**Comment by Sinan81** [comment]
After some googling and looking at cmake configs, I realized that `QSCINTILLA_LIBRARY` cmake variable should point to the library itself but not the parent folder only:
```
/disk/software/lib/hpcpm/spack/opt/spack/linux-centos7-x86_64/gcc-7.4.0/qscintilla-2.10.2-bdxsvd7op5vk2d6pnf6baghbsssghhyz/lib/libqscintilla2_qt5.so
```
it seems this took care of the above mentioned cmake warning. I am pretty sure compilation will go beyon 61% this time!

**Comment by Sinan81** [comment]
Now QGIS **build completes 71%**, next build error is:
```
              /mdal_utils.cpp                                                              │ 1 file changed, 1 insertion(+), 1 deletion(-)
     11951    CMakeFiles/spatialiteprovider.dir/qgsspatialiteprovider.cpp.o: In            │sbulut@ws-067 /mnt/local/spackdev (package/qgis) 
               function `QgsSpatiaLiteProvider::checkLayerType()':                         │$ git push
  >> 11952    /cache/sbulut/spack-stage/qgis-3.8.1-d3sanusy3l7fa7br72lkqgfzaeqc            │Counting objects: 17, done.
              gvjj/spack-src/src/providers/spatialite/qgsspatialiteprovider.cpp            │Delta compression using up to 12 threads.
              :4691: undefined reference to `sqlite3_column_table_name'                    │Compressing objects: 100% (7/7), done.
  >> 11953    collect2: error: ld returned 1 exit status                                   │Writing objects: 100% (9/9), 804 bytes | 0 bytes/s, done.
  >> 11954    make[2]: *** [output/lib/qgis/plugins/libspatialiteprovider.so] E            │Total 9 (delta 4), reused 0 (delta 0)
              rror 1                                                                       │remote: Resolving deltas: 100% (4/4), completed with 4 local objects.
     11955    make[2]: Leaving directory `/mnt/local/cache/sbulut/spack-stage/q            │To git@github.com:Sinan81/spack.git
              gis-3.8.1-d3sanusy3l7fa7br72lkqgfzaeqcgvjj/spack-build'                      │   eecf83a..60d43e2  package/qgis -> package/qgis
  >> 11956    make[1]: *** [src/providers/spatialite/CMakeFiles/spatialiteprovi            │sbulut@ws-067 /mnt/local/spackdev (package/qgis) 
              der.dir/all] Error 2                                                         │$ 
     11957    make[1]: *** Waiting for unfinished jobs....                                 ├─────────────────────────────────────────────────────────────────────────────────
     11958    [ 70%] Building CXX object src/providers/mdal/CMakeFiles/mdalprov            │sy3l7fa7br72lkqgfzaeqcgvjj/spack-build'
              ider.dir/__/__/__/external/mdal/mdal_data_model.cpp.o              
```

**Comment by Sinan81** [comment]
Tried changing sqlite dependency from build & link to build & link & run but the above mentioned build error is still persisting.

Also, played around with dependency options like `^gdal+sqlite ^sqlite+functions` and so on.  Still no luck.

Sent an email to QGIS developer mail list to see if anyone can help with this.

**Comment by Sinan81** [comment]
Hey @adamjstewart I got great news!
Finally, QGIS compilation is working!
Haven't heard back from developers yet. However, I was doing more googling, and realized that sqlite must be compiled with `SQLITE_ENABLE_COLUMN_METADATA=1` Will push a PR soon!

After QGIS UI starts we're getting run time warnings/errors about python modules `owslib` and `yaml` missing. So, I will also add these as runtime dependencies soon. One can ignore this warnings, and QGIS just runs!

**Comment by Sinan81** [comment]
Ideally, QGIS should be compiled with 5.11.1 or later to avoid a bug mentioned here: https://github.com/GNS3/gns3-gui/issues/2819#issuecomment-525656915

**Comment by Sinan81** [comment]
current install formula is:
```
spack install qgis ^gdal@2.4.2 ^ncurses+termlib ^mesa~llvm ^proj@5.2.0 ^libgeotiff@1.4.3 ^netcdf-c+mpi ^py-pyproj@1.9.6 ^python@3.7.4 ^qt@5.12.5
```
Above, `mesa~llvm` is optional, it enables to avoid building llvm which takes a loooong time.

**Comment by Sinan81** [comment]
latest version, 3.12.0 is confirmed to build successfully.

**Comment by Sinan81** [comment]
version 3.10.3 compiles just fine with Qt@5.13 as opposed to v3.8.3

**Comment by Sinan81** [comment]
confirmed that LTR versions 3.10.3 compiles with latest qt version 5.14.1 available.

**Comment by Sinan81** [comment]
Hey @adamjstewart ! I think its about time this PR is merged! The package file is in very good shape right now. I think the remaining few flake8 warning can/should be ignored. Let me know what you think.

**Comment by alalazo** [comment]
> I think the remaining few flake8 warning can/should be ignored. Let me know what you think.

Is there a reason not to fix them too?

**Comment by Sinan81** [comment]
Code would look less readable and ugly I think

**Comment by alalazo** [comment]
> Code would look less readable and ugly I think

But pleasing `flake8` makes everybody equally unhappy :wink:

File: var/spack/repos/builtin/packages/qgis/package.py [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+    # Conflicts for newer versions
+    conflicts('proj@:4.9.2', when='@3.8.2:')
+
+    # v3.8.1, Qt >= 5.9.0 is required
+    conflicts('qt@:5.8.99', when='@3.8.1:')
+    conflicts('qt@5.13.0:', when='@3.8.1')
+
+    # conflicts for qgis@2, qt@4, python@2
+    conflicts('qtkeychain@0.6.0:', when='^qt@4')
+    conflicts('qt@5:', when='@2')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/qgis/package.py]: I think a lot of these conflicts could be reworked to be `depends_on` statements. Currently, `conflicts` doesn't influence the concretizer, it just causes it to fail.

File: var/spack/repos/builtin/packages/qgis/package.py [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+
+    def cmake_args(self):
+        spec = self.spec
+        args = []
+        # qtwebkit module was removed from qt as of version 5.6
+        # needs to be compiled as a separate package
+        args.extend(['-DUSE_OPENCL=OFF',
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/qgis/package.py]:
```suggestion
        args.extend([
            '-DUSE_OPENCL=OFF',
```

File: var/spack/repos/builtin/packages/qgis/package.py [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+            # cmake couldn't determine the following paths
+            '-DEXPAT_LIBRARY={0}'.format(self.spec['expat'].libs),
+            '-DPOSTGRESQL_PREFIX={0}'.format(self.spec['postgresql'].prefix),
+            '-DQSCINTILLA_INCLUDE_DIR=' +
+                str(self.spec['qscintilla'].prefix) + '/include',
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/qgis/package.py]:
```suggestion
                self.spec['qscintilla'].prefix.include,
```

File: var/spack/repos/builtin/packages/qgis/package.py [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+            '-DQSCINTILLA_LIBRARY=' + str(self.spec['qscintilla'].prefix) +
+                '/lib/libqscintilla2_qt5.so',
+            '-DLIBZIP_INCLUDE_DIR=' +
+                str(self.spec['libzip'].prefix) + '/include',
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/qgis/package.py]:
```suggestion
                self.spec['libzip'].prefix.include,
```

File: var/spack/repos/builtin/packages/qgis/package.py [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+            '-DLIBZIP_CONF_INCLUDE_DIR=' +
+                str(self.spec['libzip'].prefix) + '/lib/libzip/include',
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/qgis/package.py]:
```suggestion
                self.spec['libzip'].prefix.lib.libzip.include,
```

File: var/spack/repos/builtin/packages/qgis/package.py [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+            '-DGDAL_CONFIG_PREFER_PATH=' + str(self.spec['gdal'].prefix.bin),
+            '-DGEOS_CONFIG_PREFER_PATH=' + str(self.spec['geos'].prefix.bin),
+            '-DGSL_CONFIG_PREFER_PATH=' + str(self.spec['gsl'].prefix.bin),
+            '-DPOSTGRES_CONFIG_PREFER_PATH=' +
+                str(self.spec['postgresql'].prefix.bin)
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/qgis/package.py]: All of these `Prefix` objects already subclass `str`, so you shouldn't need to cast them to `str`

File: var/spack/repos/builtin/packages/qgis/package.py [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+            '-DWITH_STAGED_PLUGINS={0}'.format(
+                'TRUE' if '+staged_plugin' in spec else 'FALSE'),
+            '-DWITH_THREAD_LOCAL={0}'.format(
+                'TRUE' if '+thread_local' in spec else 'FALSE'),
+            '-DWITH_TXT2TAGS_PDF={0}'.format(
+                'TRUE' if '+txt2tags_pdf' in spec else 'FALSE'),
+        ])
+
+        if '+grass7' in self.spec:
+            args.extend(['-DWITH_GRASS7=ON',
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/qgis/package.py]:
```suggestion
            args.extend([
                '-DWITH_GRASS7=ON',
```

File: var/spack/repos/builtin/packages/qgis/package.py (hunk 6 above)
Comment (Sinan81) [reply var/spack/repos/builtin/packages/qgis/package.py]: Good to know.

File: var/spack/repos/builtin/packages/qgis/package.py [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+        args = []
+        # qtwebkit module was removed from qt as of version 5.6
+        # needs to be compiled as a separate package
+        args.extend([
+            '-DUSE_OPENCL=OFF',
+            # cmake couldn't determine the following paths
+            '-DEXPAT_LIBRARY={0}'.format(self.spec['expat'].libs),
+            '-DPOSTGRESQL_PREFIX={0}'.format(self.spec['postgresql'].prefix),
+            '-DQSCINTILLA_INCLUDE_DIR=' +
+                self.spec['qscintilla'].prefix.include,
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/qgis/package.py]:
```
var/spack/repos/builtin/packages/qgis/package.py:127: [E131] continuation line unaligned for hanging indent
var/spack/repos/builtin/packages/qgis/package.py:129: [E131] continuation line unaligned for hanging indent
```
Try no additional indentation.

Comment (Sinan81) [reply var/spack/repos/builtin/packages/qgis/package.py]: no flake8 errors when I check it on my laptop :confused:

Comment (Sinan81) [reply var/spack/repos/builtin/packages/qgis/package.py]: I think that is fixed as of last commit.

**Comment by adamjstewart** [comment]
If you install `flake8` you can run `spack flake8` locally.

**Comment by Sinan81** [comment]
> If you install `flake8` you can run `spack flake8` locally.

I already do that. Oddly enough `spack flake8` checks are fully clean right now on my end

**Comment by Sinan81** [comment]
just did `spack flake8` on another machine, it still looks fully clean.

**Comment by Sinan81** [comment]
:tada:

**Comment by Sinan81** [comment]
Hey @pat-s This PR might be of interest to you!

**Comment by Sinan81** [comment]
Just found a minor typo in the package. And also realized something slightly changed last time I build the package, and build fails around 100% build progress. Will fix these soon

**Comment by Sinan81** [comment]
Update: I think the build fail was due to a typo. Will push a PR soon.

**Comment by Sinan81** [comment]
Above PR fixes the typo and the build issue.

**Pull request merged**